Orange/widgets/model/owneuralnetwork.py

```diff
@@ -142,35 +156,42 @@ def __update(self):
             self.cancel()
         assert self._task is None
 
-        self.setBlocking(True)
-
-        self._task = task = Task()
-
-        # A thread safe way to invoke a method
-        set_progress = methodinvoke(self, "setProgressValue", (float,))
-
         max_iter = self.learner.kwargs["max_iter"]
 
+        # Setup the task state
+        task = Task()
+        lastemitted = 0.
+
         def callback(iteration):
-            if task.cancelled:
-                raise CancelThreadException()  # this stop the thread
-            set_progress(iteration/max_iter*100)
+            nonlocal task  # type: Task
+            nonlocal lastemitted
+            if task.isInterruptionRequested():
+                raise CancelTaskException()
+            progress = round(iteration / max_iter * 100)
+            if progress != lastemitted:
+                task.emitProgressUpdate(progress)
+                lastemitted = progress
 
         # copy to set the callback so that the learner output is not modified
         # (currently we can not pass callbacks to learners __call__)
         learner = copy.copy(self.learner)
         learner.callback = callback
 
         def build_model(data, learner):
-            return learner(data)
+            try:
+                return learner(data)
+            except CancelTaskException:
```

Comment:
What is the difference between former (no try-except) and this?

Reply:
The later does not capture the exception instance's `__traceback__` in the resulting Future (which we know will be discarded).